src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/evolutionaryAlgorithm/CostModel.java

```diff
@@ -0,0 +1,304 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.evolutionaryAlgorithm;
+
+import se.liu.ida.hefquin.engine.federation.*;
+import se.liu.ida.hefquin.engine.federation.access.DataRetrievalRequest;
+import se.liu.ida.hefquin.engine.federation.access.SPARQLRequest;
+import se.liu.ida.hefquin.engine.federation.access.TriplePatternRequest;
+import se.liu.ida.hefquin.engine.query.BGP;
+import se.liu.ida.hefquin.engine.query.TriplePattern;
+import se.liu.ida.hefquin.engine.queryplan.LogicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.PhysicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpBGPAdd;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpJoin;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpRequest;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpTPAdd;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalOperatorForLogicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.*;
+import se.liu.ida.hefquin.engine.queryplan.utils.ExpectedVariablesUtils;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+import se.liu.ida.hefquin.engine.queryproc.QueryProcContext;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.CardinalityEstimation;
+
+import java.util.*;
+
+public class CostModel extends CardinalityEstimation{
+    protected final QueryProcContext ctxt;
+    protected final RootOperatorCostCache rootOperatorCostCache = new RootOperatorCostCache();
+
+    public CostModel(QueryProcContext ctxt) {
+        super(ctxt);
+        this.ctxt = ctxt;
+    }
+
+    public Double overallCost( final PhysicalPlan pp ) throws QueryOptimizationException {
+        List<Integer> measureMetrics = costMetrics( pp, new ArrayList<Integer>() );
+        ArrayList<Double> weight= new ArrayList<Double>(Arrays.asList(0.2, 0.2, 0.2, 0.2, 0.2));
+
+        double overallCost = 0;
+        for (  int i = 0; i < measureMetrics.size(); i++ ){
+            overallCost = overallCost + measureMetrics.get(i) * weight.get(i);
+        }
+
+        return overallCost;
+    }
+
+    protected List<Integer> costMetrics( final PhysicalPlan pp, final List<Integer> overallMetrics ) throws QueryOptimizationException {
```

Comment:
This return type is totally meaningless. Please introduce an actual interface for the objects that this method returns, where this interface has a specific getter method for each of the five metrics, e.g., getNumberOfRequests().